File: app/catalog.py
"""Scan upstream + overlay into categories and modules for the UI."""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path

from . import config, overlay
from .sysmon_xml import parse_file

_cache: dict[str, tuple[float, "Module"]] = {}

CAT_PREFIX_RE = re.compile(r"^(\d+(?:_\d+)*)_(.+)$")


@dataclass
class Module:
    rel: str
    category: str
    filename: str
    kind: str                     # include | exclude | other
    source: str                   # upstream | edited | custom
    title: str                    # first RuleGroup name or filename
    event_types: list[str] = field(default_factory=list)
    techniques: list[tuple[str, str]] = field(default_factory=list)
    rule_count: int = 0
    error: str = ""


@dataclass
class Category:
    dirname: str
    event_ids: list[str]
    label: str
    modules: list[Module] = field(default_factory=list)

    @property
    def include_count(self) -> int:
        return sum(1 for m in self.modules if m.kind == "include")

    @property
    def exclude_count(self) -> int:
        return sum(1 for m in self.modules if m.kind == "exclude")


def category_label(dirname: str) -> tuple[list[str], str]:
    m = CAT_PREFIX_RE.match(dirname)
    if not m:
        return [], dirname
    ids = m.group(1).split("_")
    label = m.group(2).replace("_", " ").capitalize()
    return ids, label


def filename_title(fname: str) -> str:
    """exclude_adobe_acrobat.xml -> 'Adobe acrobat'"""
    stem = re.sub(r"^(include|exclude)_", "", fname.removesuffix(".xml"))
    return stem.replace("_", " ").replace("-", " ").strip().capitalize() or fname


def _sort_key(dirname: str) -> tuple[int, str]:
    m = CAT_PREFIX_RE.match(dirname)
    return (int(m.group(1).split("_")[0]) if m else 999, dirname)


def parsed_module(rel: str):
    """Parsed sysmon_xml.Module for a module path, cached on file mtime."""
    path = overlay.resolve(rel)
    mtime = path.stat().st_mtime
    cached = _cache.get(str(path))
    if cached and cached[0] == mtime:
        return cached[1]
    parsed = parse_file(path)
    _cache[str(path)] = (mtime, parsed)
    return parsed


def _load_module(rel: str) -> Module:
    cat, fname = rel.split("/")
    path = overlay.resolve(rel)
    kind = "include" if fname.startswith("include_") else "exclude" if fname.startswith("exclude_") else "other"
    mod = Module(rel=rel, category=cat, filename=fname, kind=kind, source=overlay.source_of(rel), title=fname)
    try:
        parsed = parsed_module(rel)
        mod.title = next((rg.name for rg in parsed.rulegroups if rg.name), "") or filename_title(fname)
        mod.event_types = parsed.event_types
        mod.techniques = parsed.techniques
        mod.rule_count = parsed.rule_count
    except Exception as exc:  # broken XML in overlay must not break the page
        mod.error = str(exc)
    return mod


def category_dirs() -> list[str]:
    names = {p.name for p in config.UPSTREAM_DIR.iterdir() if p.is_dir() and overlay.CATEGORY_RE.match(p.name)}
    if config.OVERLAY_DIR.exists():
        names |= {p.name for p in config.OVERLAY_DIR.iterdir() if p.is_dir() and overlay.CATEGORY_RE.match(p.name)}
    return sorted(names, key=_sort_key)


def module_rels(category: str) -> list[str]:
    files: set[str] = set()
    for base in (config.UPSTREAM_DIR / category, config.OVERLAY_DIR / category):
        if base.is_dir():
            files |= {p.name for p in base.glob("*.xml")}
    return [f"{category}/{f}" for f in sorted(files)]


def scan() -> list[Category]:
    cats: list[Category] = []
    for d in category_dirs():
        ids, label = category_label(d)
        cat = Category(dirname=d, event_ids=ids, label=label)
        cat.modules = [_load_module(rel) for rel in module_rels(d)]
        if cat.modules:
            cats.append(cat)
    return cats


def all_module_rels() -> list[str]:
    return [rel for d in category_dirs() for rel in module_rels(d)]


def get_category(dirname: str) -> Category | None:
    return next((c for c in scan() if c.dirname == dirname), None)


def get_module(rel: str) -> Module:
    return _load_module(rel)

File: site/js/catalog.js
// Effective module catalog = upstream catalog.json + the user's overlay (edited/custom modules).
import { parseModule, summarize } from "./model.js";

export const CATEGORY_RE = /^\d+(?:_\d+)*_[a-z0-9_]+$/;
const CAT_PREFIX_RE = /^(\d+(?:_\d+)*)_(.+)$/;

export function categoryMeta(dirname) {
  const m = CAT_PREFIX_RE.exec(dirname);
  return { dirname, event_ids: m ? m[1].split("_") : [], label: m ? m[2].replace(/_/g, " ").replace(/^./, c => c.toUpperCase()) : dirname };
}
const sortKey = d => { const m = CAT_PREFIX_RE.exec(d); return m ? parseInt(m[1].split("_")[0], 10) : 999; };

export function kindOf(rel) {
  const f = rel.split("/")[1] ?? "";
  return f.startsWith("include_") ? "include" : f.startsWith("exclude_") ? "exclude" : "other";
}
export function filenameTitle(fname) {
  const stem = fname.replace(/\.xml$/, "").replace(/^(include|exclude)_/, "");
  return stem.replace(/[_-]+/g, " ").trim().replace(/^./, c => c.toUpperCase()) || fname;
}
export function validRel(rel) {
  const parts = (rel || "").split("/");
  return parts.length === 2 && CATEGORY_RE.test(parts[0]) && parts[1].endsWith(".xml") && !rel.includes("..");
}
export function slugify(name) {
  const s = name.trim().toLowerCase().replace(/[^a-z0-9]+/g, "_").replace(/^_+|_+$/g, "");
  if (!s) throw new Error("name is empty");
  return s;
}

const summaryCache = new Map(); // xml → summary (modules are immutable strings)
function summaryFor(xml) {
  let s = summaryCache.get(xml);
  if (!s) {
    try { s = { ...summarize(parseModule(xml)), error: "" }; }
    catch (e) { s = { title: "", event_types: [], techniques: [], rule_count: 0, error: e.message }; }
    if (summaryCache.size > 2000) summaryCache.clear();
    summaryCache.set(xml, s);
  }
  return s;
}

export class Catalog {
  constructor(data, overlay) {
    this.upstream = new Map(data.modules.map(m => [m.rel, m.xml]));
    this.template = data.template || "";
    this.examples = data.examples || {};
    this.presets = data.presets || [];
    this.mdeCovered = data.mde_covered || [];
    this.generated = data.generated;
    this.upstreamCategories = data.categories.map(c => c.dirname);
    this.overlay = overlay; // live object {rel: xml}
  }
  exists(rel) { return this.overlay[rel] !== undefined || this.upstream.has(rel); }
  xml(rel) { return this.overlay[rel] ?? this.upstream.get(rel); }
  sourceOf(rel) { return this.overlay[rel] !== undefined ? (this.upstream.has(rel) ? "edited" : "custom") : "upstream"; }
  categoryDirs() {
    const set = new Set(this.upstreamCategories);
    for (const rel of Object.keys(this.overlay)) set.add(rel.split("/")[0]);
    return [...set].filter(d => CATEGORY_RE.test(d)).sort((a, b) => sortKey(a) - sortKey(b) || a.localeCompare(b));
  }
  moduleRels(category) {
    const files = new Set();
    for (const rel of this.upstream.keys()) if (rel.startsWith(category + "/")) files.add(rel);
    for (const rel of Object.keys(this.overlay)) if (rel.startsWith(category + "/")) files.add(rel);
    return [...files].sort();
  }
  allRels() { return this.categoryDirs().flatMap(d => this.moduleRels(d)); }
  module(rel) {
    const xml = this.xml(rel);
    if (xml === undefined) return null;
    const [category, filename] = rel.split("/");
    const s = summaryFor(xml);
    return { rel, category, filename, kind: kindOf(rel), source: this.sourceOf(rel), ...s, title: s.title || filenameTitle(filename) };
  }
  parsed(rel) { return parseModule(this.xml(rel)); }
  categories() {
    return this.categoryDirs().map(d => ({ ...categoryMeta(d), modules: this.moduleRels(d).map(r => this.module(r)) })).filter(c => c.modules.length);
  }
  overlayRels() { return Object.keys(this.overlay).filter(validRel).sort(); }
}
